### ISSUE TITLE
Regression in 0.6.18 with dictionary conditionals

### ISSUE THREAD

**Issue by vinayinvicible** [Nuitka/Nuitka#1348]
* Nuitka version, full Python version, flavor, OS, etc. as output by this
  command (it does more than you think, and we are adding more all the time):

  > python -m nuitka --version
0.6.18
Commercial: None
Python: 3.9.5 (default, May 10 2021, 20:23:31) 
Flavor: pyenv
Executable: ~/.virtualenvs/nuitka-test/bin/python
OS: Linux
Arch: x86_64
Distribution: Ubuntu "20.04.3 LTS (Focal Fossa)"

* How did you install Nuitka and Python

  > Installed Python using pyenv.
     Installed Nuitka via pip. 

* Provide in your issue the Nuitka options used

  ```python
  #   test.py
    def main():
        params = {}
        if True:
            params.update({"abc": "asd"})

        # call all senders at once
        status = 0
        if params:
            status = 1
        return status


    if __name__ == "__main__":
        print(main())
  ```
  `python -m nuitka test.py --onefile -o test`

  ```bash
  >> python test.py
  1
  >> ./test
  0
  ```

* Note if this is a regression

  binary output is correct if
  * it is generated using 0.6.17.7
  * dictionary in line 2 is declared using builtin `dict()`

**Comment by kayhayen** [comment]
So, this one is to blame, which I believe is a bit early, I was thinking a mistake would be made by the dictionary update operation:

```
commit 0bbfbb5bbf9d9a47a8dd2660ca2ff1757193ce31
Author: Kay Hayen <kay.hayen@gmail.com>
Date:   Thu Oct 28 12:46:46 2021 +0200

    Optimization: Add attribute nodes for all dict attributes
    
    * These are added without any specific optimization yet.
    
    * The operations are to be filled out later, and will be
      automatically detected by the generator.

```

**Comment by kayhayen** [comment]
The error is though, that the condition `if params` is working with a truth check, that would consider escaping, but when the transformation from fixed `update` attribute lookup is done to type specifc `update` attribute lookup, the escaping is not done for one pass, and that's enough:

```
    def computeExpression(self, trace_collection):
        subnode_expression = self.subnode_expression

        if subnode_expression.hasShapeDictionaryExact():
            result = ExpressionAttributeLookupDictUpdate(
                expression=subnode_expression, source_ref=self.source_ref
            )

            return (
                result,
                "new_expression",
                "Attribute lookup 'update' on dict shape resolved.",
            )

        return subnode_expression.computeExpressionAttribute(
            lookup_node=self,
            attribute_name="update",
            trace_collection=trace_collection,
        )
```

There is a method in trace collections, that will immediately evaluate the expression, which wasn't used here, therefore allowing the bug.

**Comment by kayhayen** [comment]
I have pushed a correction for this to factory branch: https://nuitka.net/doc/factory.html

Not sure how to deal with this, I think I want to rebase things, but the code has obviously changed many times with new ones added. But I guess I only need to use new generation template on old code config maybe.

**Comment by kayhayen** [comment]
Released as part of 0.6.18.6 hotfix
